accepted/2021/statics-in-interfaces/README.md

```diff
@@ -8,6 +8,791 @@ C# is looking at enabling static abstract members in interfaces (https://github.
+
+        static abstract bool operator >(TSelf lhs, TOther rhs);
+
+        static abstract bool operator >=(TSelf lhs, TOther rhs);
+    }
+
+    public interface IAddable<TSelf, TOther>
+        where TSelf : IAddable<TSelf, TOther>
+    {
+        // An additive identity is not exposed here as it may be problematic for cases like
+        // DateTime that implements both IAddable<DateTime, DateTime> and IAddable<DateTime, TimeSpan>
+        // This would require overloading by return type and may be problematic for some scenarios where TOther might not have an additive identity
+
+        // We can't assume TOther + TSelf is valid as not everything is commutative
+
+        // We can't expose TSelf - TOther as some types only support one of the operations
+
+        static abstract TSelf operator +(TSelf value);
+
+        static abstract TSelf operator +(TSelf lhs, TOther rhs);
+
+        // There is an open question on if we should support "checked" versions of the operators
+        // Many algorithms, such as in LINQ, depend on checked behavior for overflow
+        // Supporting this would require well-defined ECMA-335 names and language work to ensure the right things happen
+    }
+
+    public interface ISubtractable<TSelf, TOther>
+        where TSelf : ISubtractable<TSelf, TOther>
+    {
+        static abstract TSelf operator -(TSelf lhs, TOther rhs);
+    }
+
+    public interface IIncrementable<TSelf>
+        where TSelf : IIncrementable<TSelf>
+    {
+        // We don't inherit IAddable as TOther isn't well-defined for incrementable types
+        // Incrementing twice is not necessarily the same as self + 2 and not all languages
+        // use incrementing for numbers, they are sometimes used by iterators for example
+
+        // We can't expose TSelf-- as some types only support one of the operations
+
+        static abstract TSelf operator ++(TSelf value);
+    }
+
+    public interface IDecrementable<TSelf>
+        where TSelf : IDecrementable<TSelf>
+    {
+        static abstract TSelf operator --(TSelf value);
+    }
+
+    public interface IMultipliable<TSelf, TOther>
+        where TSelf : IMultipliable<TSelf, TOther>
+    {
+        // An multiplicative identity is not exposed here for the same reasons as IAddable<TSelf, TOther>
+
+        // We can't assume TOther * TSelf is valid as not everything is commutative
+
+        // We can't expose TSelf / TOther as some types, such as matrices, don't support dividing with respect to TOther
+
+        // We can't inherit from IAddable<TSelf, TOther> as some types, such as matrices, support Matrix * scalar but not Matrix + scalar
+
+        static abstract TSelf operator *(TSelf lhs, TOther rhs);
+    }
+
+    public interface IDivisable<TSelf, TOther>
+        where TSelf : IDivisable<TSelf, TOther>
+    {
+        static abstract TSelf operator /(TSelf lhs, TOther rhs);
+    }
+
+    public interface IRemainder<TSelf, TOther>
```

Comment:
Are there cases where something is "remaindable" without being divisible?

Reply:
Not that I've encountered nor that I can think of off the top of my head. So it might be possible this can inherit from `IDivisable` (C# does not require them to be paired, however, so it is possible).

Reply:
I should have also asked if the inverse is possible? Can something be divisible without supporting the concept of a remainder? I'm wondering if we could just collapse into `IDivisible` with both operators.

Reply:
Yes, there are cases like Matrix or even TimeSpan that only support division today. We could potentially extend them to also support remainder, but it seems like a restriction that might not apply in practice.

Reply:
Technically numbers that support remainders are known as Euclidean domains. Calling it IEuclidean is probably a stretch but how about IModularizable?